xarray/backends/store.py

```diff
@@ -56,8 +58,22 @@ def open_dataset(
             decode_timedelta=decode_timedelta,
         )
 
-        ds = Dataset(vars, attrs=attrs)
-        ds = ds.set_coords(coord_names.intersection(vars))
+        # split data and coordinate variables (promote dimension coordinates)
+        data_vars = {}
+        coord_vars = {}
+        for name, var in vars.items():
+            if name in coord_names or var.dims == (name,):
+                coord_vars[name] = var
+            else:
+                data_vars[name] = var
+
+        if set_indexes:
+            coords = coord_vars
+        else:
+            # explicit Coordinates object with no index passed
+            coords = Coordinates(coord_vars)
```

Comment:
we discussed this in the raster index meeting a bit with @dcherian and @benbovy, and if I remember correctly we agreed that instead of passing through an option to the backend it might be cleaner to _always_ create the backend dataset without indexes, and create the indexes in `open_dataset`:
```suggestion
        # explicit Coordinates object with no index passed
        coords = Coordinates(coord_vars)
```
Most known backends use the `BackendStoreEntrypoint` to assemble the backend dataset, so these backends will work automatically, and for all other backends we'll have to figure out a way to warn.

If we also change the data model to cleanly separate I/O from decoding variables (what `zarr` calls "ArrayToArray" codecs, and what our current decoders do) and datasets (this would be where you'd create indexes), then we could warn about backend datasets with indexes. I think this would be a breaking change, although we could add a deprecation cycle to make the transition smoother.

Reply:
I've taken the liberty of pushing directly to this PR (the failing flaky test is real but unrelated, and `pre-commit` also fails on  `main`), so the state in this PR should be closer to what I'm proposing above.

cc @dcherian, @benbovy

Reply:
I've changed my mind. In reality, we are seeing that people are writing backends not just for "reading arrays", they are creating datasets with extra metadata (e.g. rioxarray might choose to add RasterIndex). So perhaps we add a `create_default_indexes` that sets the default PandasIndex for any dimension coordinate that does not have an index, when returned from the backend?

Reply:
if you look at the current state of the PR, that's exactly my intention

Reply:
> So perhaps we add a `create_default_indexes` that sets the default PandasIndex for any dimension coordinate that does not have an index, when returned from the backend?

The general version of this would be to make the index-adding step independently pluggable, i.e. pass in a callable, which defaults to your `create_default_indexes`

Reply:
That makes sense, but then do we deprecate the kwarg you're about to add?

Reply:
Pluggable and composable Dataset (de)coders would be nice indeed for setting indexes among other things. I like the proposal made in #8548.

Even with a decoder API it might still make sense to have a special case (kwarg) for default indexes, though, since those are also created in `Dataset.__init__()`, `DataArray.__init__()`, `Coordinates.__init__()`, `assign_coords()`, etc.